Add Withings support

vangorra opened the pull request:
Reopening #21625

## Description:
Adding support for connecting to Withings apis to integrate health data into home assistant.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8816

## Example entry for `configuration.yaml` (if applicable):
```yaml
# Example configuration.yaml entry
withings:
    client_id: <client id from previous step>
    client_secret: <consumer secret from previous step>
    profiles: 
        - <name of user profile>
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) [PR](https://github.com/home-assistant/home-assistant.io/pull/8816)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

balloob commented:
Did you address my comments from the previous PR?

vangorra commented:
Yes. The config flow changes were especially  welcomed as it simplified the code.

MartinHjelmare commented on homeassistant/components/withings/common.py:
```diff
@@ -247,3 +249,38 @@ def function():
+
+    _LOGGER.debug(
+        "Creating nokia api instance with credentials %s",
+        credentials
```
We shouldn't log credentials.

MartinHjelmare commented on homeassistant/components/withings/common.py:
```diff
@@ -0,0 +1,286 @@
+    credentials.__dict__.update(entry.data[const.CREDENTIALS])
+
+    def credentials_saver(credentials_param):
+        _LOGGER.debug("Saving updated credentials.")
```
We don't end logging messages with period.

MartinHjelmare commented on homeassistant/components/withings/config_flow.py:
```diff
@@ -0,0 +1,209 @@
+        """Create an entry for selecting a profile."""
+        flow = self.hass.data.get(DATA_FLOW_IMPL, {})
+
+        if user_input and const.PROFILE in user_input:
```
The profile is required in the form schema so we shouldn't need to check for that here.

vangorra replied:
This is necessary as to short-circuit the config flow to goto the auth flow if the profile is provided.

MartinHjelmare replied:
The profile will always be provided if any user input is provided.

vangorra replied:
I'm a bit confused on what's being requested here. The check for the profile is necessary to advance the flow to the "auth" step. This may be based on a lack of understanding config flows. This is how I understand the config flow process to work:
- Config flow gets "user" step, which returns a show form step with the profile list.
- User selects a profile from the list.
- Config flow gets "user" step with the user's input. Which returns the external step with id of "auth".

MartinHjelmare replied:
This is enough.
```py
if user_input:
```

MartinHjelmare commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,427 @@
+
+        if isinstance(self._attribute, WithingsMeasureAttribute):
+            _LOGGER.debug("Updating measures state")
+            self._data_manager.update_measures()
```
The nokia library doesn't support asyncio, it seems. Is this call doing I/O? We're not allowed to do I/O in the event loop.

MartinHjelmare commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,427 @@
+                self._data_manager.sleep_summary
+            )
+
+    async def async_update_measure(self, data) -> None:
```
There's no `await` in this method, so it doesn't need to be a coroutine. Please define functions that don't use await as regular callbacks. Those have less overhead.

homeassistant commented:
Hello @vangorra,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`665c5ebb73c645c2b35141e1256f28b6df04ceb6`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;665c5ebb73c645c2b35141e1256f28b6df04ceb6) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`b2478a1b0cdc5390db0136e7a414bae6a8ff80e7`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;b2478a1b0cdc5390db0136e7a414bae6a8ff80e7) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`9c02e0998cadcce116adc60bd7d6d5307273841d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;9c02e0998cadcce116adc60bd7d6d5307273841d) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`d6e8b3422a2573bc423b6817e93ea4298cceca91`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;d6e8b3422a2573bc423b6817e93ea4298cceca91) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`e5021de18110f4797c560166a65df76c10335cbd`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;e5021de18110f4797c560166a65df76c10335cbd) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`f741fb8c9da4211b699a4dc21201c95d3753b6ac`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;f741fb8c9da4211b699a4dc21201c95d3753b6ac) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`4918653e69abe1b4df9f09143e09003612f70c2d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;4918653e69abe1b4df9f09143e09003612f70c2d) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`fa367646e772bb8cd617d04add1d0f4390d9e078`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;fa367646e772bb8cd617d04add1d0f4390d9e078) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`1af0f10c30566f4337de6980029bcf8ed05b0e6b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;1af0f10c30566f4337de6980029bcf8ed05b0e6b) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`ea7d7fa4a3529b2fb4a32b14122da620536f1511`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;ea7d7fa4a3529b2fb4a32b14122da620536f1511) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`46b1738d7173a57213307c93e24c72e211fd6ec6`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;46b1738d7173a57213307c93e24c72e211fd6ec6) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`075a57a764ab347c388a50b1a3e088f2253a6185`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;075a57a764ab347c388a50b1a3e088f2253a6185) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`c211d5949790d4cacab03c1c7288b79a00f4c018`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;c211d5949790d4cacab03c1c7288b79a00f4c018) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`67d934b06adeab933546683dfbb9fa3322e3eaaf`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;67d934b06adeab933546683dfbb9fa3322e3eaaf) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`2515d6f263799dacfd40445387d29692dda9a4a2`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;2515d6f263799dacfd40445387d29692dda9a4a2) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`cef3af5c678ad449e58d916e5a1fa0cf835fa0bf`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;cef3af5c678ad449e58d916e5a1fa0cf835fa0bf) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`9f598e254d62962127d3920a0b9eea9c4cfbed63`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;9f598e254d62962127d3920a0b9eea9c4cfbed63) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`927389af60f6da2942e974715caf380243bdad4b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;927389af60f6da2942e974715caf380243bdad4b) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`8beb1524d1ab77f21b73b6464b2a6da10cc56b1d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;8beb1524d1ab77f21b73b6464b2a6da10cc56b1d) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`d22e92f4afe1191e6f4d5fd7c8a8526d10aaf414`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;d22e92f4afe1191e6f4d5fd7c8a8526d10aaf414) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`db61b2acebba97ee6e80dc3ca171e784c28bad75`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;db61b2acebba97ee6e80dc3ca171e784c28bad75) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`8248822725c566bd9baeb86dfe55f541f92056e3`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;8248822725c566bd9baeb86dfe55f541f92056e3) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`a43c4cbb29174845a4d9a1f4aa7cac1a250a3d86`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;a43c4cbb29174845a4d9a1f4aa7cac1a250a3d86) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`553f038d934016af473fb8a2f7c568fdcd5b4824`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;553f038d934016af473fb8a2f7c568fdcd5b4824) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`4559d0e8780c74d9e808bbdecd89691ccb8cfb75`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;4559d0e8780c74d9e808bbdecd89691ccb8cfb75) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`1860e5330270bdff4e7fa9ab5c3109ec352c78cc`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;1860e5330270bdff4e7fa9ab5c3109ec352c78cc) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`b5863cb87d860dd0c5582a314aa4fd9c961baf60`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;b5863cb87d860dd0c5582a314aa4fd9c961baf60) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`c7327e02bb7a549f2bf55e8e0003272033379253`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;c7327e02bb7a549f2bf55e8e0003272033379253) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`498b628e54ff9f2d0004fa5b789e6797c8597edc`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;498b628e54ff9f2d0004fa5b789e6797c8597edc) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`aa4b0f16c07fd0937dd0f938261d9bc4f02b1ed8`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;aa4b0f16c07fd0937dd0f938261d9bc4f02b1ed8) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`d29a1b17ba5b16d8b1f1306e0450ccc374104eb0`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;d29a1b17ba5b16d8b1f1306e0450ccc374104eb0) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`4a7c29d995f0b9804138da2160040fb403c0bea3`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;4a7c29d995f0b9804138da2160040fb403c0bea3) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`79fe2eed55b2b9378531ab96ae463644b6229722`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;79fe2eed55b2b9378531ab96ae463644b6229722) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`6bf79ca07f73a879e8d44f5157882c608c78913e`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;6bf79ca07f73a879e8d44f5157882c608c78913e) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`f62697d34edd73fb375e8fa9ced9b2e04396bac6`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;f62697d34edd73fb375e8fa9ced9b2e04396bac6) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`8a16ce6f9ae9c8f6d3dabb73bd1216138e45beb5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;8a16ce6f9ae9c8f6d3dabb73bd1216138e45beb5) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`d922357a8486a837a99709a8967339705fa7a627`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;d922357a8486a837a99709a8967339705fa7a627) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`73e801caddae8e49a7bf1b35af30bf262603c1e4`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;73e801caddae8e49a7bf1b35af30bf262603c1e4) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`0066d268f0965e0d0d2112bbd31e96664171a4fb`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;0066d268f0965e0d0d2112bbd31e96664171a4fb) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`c396dff333f4f03c48e38609b24bd2fc3145edec`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;c396dff333f4f03c48e38609b24bd2fc3145edec) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`7fa353b25e1b71af055a267bbc1decaf0415e848`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;7fa353b25e1b71af055a267bbc1decaf0415e848) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`0563715570054fcc07a22ad6db6764c841298f52`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;0563715570054fcc07a22ad6db6764c841298f52) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`bd2ca52b8510c361484592259291f566cb9adfa2`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;bd2ca52b8510c361484592259291f566cb9adfa2) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`e7ecbeb197347c2987efab8e4f6e01598489e259`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;e7ecbeb197347c2987efab8e4f6e01598489e259) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`1fc98df2a4e6958d92eb8d4863d49f355685b084`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;1fc98df2a4e6958d92eb8d4863d49f355685b084) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`3d57cdaa72ad3eaba5365d4933b9802d36ff12fa`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;3d57cdaa72ad3eaba5365d4933b9802d36ff12fa) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`c1c4df20406bb39facd775d8f9f9fcae49a66b96`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;c1c4df20406bb39facd775d8f9f9fcae49a66b96) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`867896da542c0ed90adbabc110725e552cf04dda`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;867896da542c0ed90adbabc110725e552cf04dda) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`c695b711f542204423cdd178eeb60b86881e8d7b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;c695b711f542204423cdd178eeb60b86881e8d7b) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`4efcf4be1fb0da49c1ecec46a9d1658aae1bb684`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;4efcf4be1fb0da49c1ecec46a9d1658aae1bb684) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`3e364458d0b9a8d0fa455e3a3a603d088b460d8b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;3e364458d0b9a8d0fa455e3a3a603d088b460d8b) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`799a3e8820fd350d8e9e15df627974e411555ff0`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;799a3e8820fd350d8e9e15df627974e411555ff0) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`54389e985ebfb56641b6df19f6606debc262ab44`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;54389e985ebfb56641b6df19f6606debc262ab44) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`9c70e516ac095ea922ad05c3c1dd0a9c03deaa26`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;9c70e516ac095ea922ad05c3c1dd0a9c03deaa26) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`2f2bbb774743f9b4ff1756bb5734ff3488e024e1`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;2f2bbb774743f9b4ff1756bb5734ff3488e024e1) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`67e20b5223cc559fcc9cbe6f603bd62a18c3e0bc`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;67e20b5223cc559fcc9cbe6f603bd62a18c3e0bc) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`1e6856b59bd760b7c1b704fca4f9fc8a4ef05173`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;1e6856b59bd760b7c1b704fca4f9fc8a4ef05173) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`cead180ad6296e6cf204ec85ffc81324c31cbc47`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;cead180ad6296e6cf204ec85ffc81324c31cbc47) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`abe5391dc95f958a6823a15afa93fa0a6410f7dd`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;abe5391dc95f958a6823a15afa93fa0a6410f7dd) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`a7ede5cdee3287fd060728842998e57bdcd11400`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;a7ede5cdee3287fd060728842998e57bdcd11400) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`db671a6f0dc2f505e9aa3a1953bdae84eb1c374a`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;db671a6f0dc2f505e9aa3a1953bdae84eb1c374a) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`0b6662f8d9ff791bc643555d0baa89624ddd48fc`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;0b6662f8d9ff791bc643555d0baa89624ddd48fc) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`ecc09c103889b2c5b54d7f6d8aeb70ba8d09fd83`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;ecc09c103889b2c5b54d7f6d8aeb70ba8d09fd83) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`0adc0375efff71795a3add14ecda97e97dd85900`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;0adc0375efff71795a3add14ecda97e97dd85900) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`00d71afb8b1382c9efef18a2234ef08da79abb47`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;00d71afb8b1382c9efef18a2234ef08da79abb47) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`4b9b59fbb2242e5b75fc1c1f7a75bae65ab3aedc`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;4b9b59fbb2242e5b75fc1c1f7a75bae65ab3aedc) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`110fb9ab0711edd574edcd22ace11ad63a36b93f`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;110fb9ab0711edd574edcd22ace11ad63a36b93f) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.
- [`6c235c4fc776dc723bd670324661e81341ea1c20`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;6c235c4fc776dc723bd670324661e81341ea1c20) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hello @vangorra,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`6cc9786d055bdc4bbf75ecc7bff60b8d8cfaa44e`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25154&#x2F;commits&#x2F;6cc9786d055bdc4bbf75ecc7bff60b8d8cfaa44e) - This commit has something that looks like an email address (robbie.vangorkom@gmail.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

vangorra commented:
I didn't realize changing my github email would be so messy. At any rate the PR is updated and addresses the latest feedback.

MartinHjelmare commented:
The branch isn't clean. It has lots of unrelated commits. Please rebase and clean the branch, or open a new PR from a clean branch.

vangorra committed:
Rebasing with a clean branch.
Addressing PR feedback.
Cleaning up some static code checks.
Fixing bug with saving credentials.

vangorra committed:
Removing unecessary change.

MartinHjelmare commented on homeassistant/components/withings/config_flow.py:
```diff
@@ -0,0 +1,184 @@
+def register_flow_implementation(hass, client_id, client_secret, base_url, profiles):
+    """Register a flow implementation.
+
+    domain: Domain of the component responsible for the implementation.
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,454 @@
+    async_add_entities: types.Callable[[types.List[Entity], bool], None],
+):
+    """Set up the sensor config entry."""
+    data_manager = create_withings_data_manger(hass, entry)
```
If we would store the data manager instance in hass.data we wouldn't need to create it again here after forwarding the config entry.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,112 @@
+            )
+        )
+        return False
+
```
We can store the data manager in hass.data here.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,112 @@
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": const.SOURCE_USER}, data={}
```
This should be `SOURCE_IMPORT`.

vangorra replied:
Does that mean 'async_step_user' should also be renamed to 'async_step_import'?

MartinHjelmare replied:
No, the user and import steps are different entry points to the config flow. It is common to pass the data from the import step directly to the user step, if no extra validation is required in the import step.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,112 @@
+                    [vol.All(cv.string, vol.Length(min=1))],
+                ),
+                vol.Optional(
+                    const.MEASURES, default=list(WITHINGS_MEASUREMENTS_MAP)
```
All measurements should be created without an option to select them. See:
https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md

vangorra committed:
Caching data manager.

vangorra committed:
Removing configurable measures.

vangorra committed:
Using import step in config flow.

vangorra committed:
Updating config flows.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,103 @@
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": const.SOURCE_IMPORT}, data={}
```
Please use the source constants from config_entries.py:
https://github.com/home-assistant/home-assistant/blob/ca74a23cf191dc6ca081d1496c9d0afbbfd65b04/homeassistant/config_entries.py#L27-L29

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,103 @@
+            hass.config_entries.flow.async_init(
+                const.DOMAIN,
+                context={
+                    "source": const.SOURCE_USER,
```
See above.

MartinHjelmare commented on homeassistant/components/withings/common.py:
```diff
@@ -0,0 +1,308 @@
+        return self._sleep_summary
+
+
+def create_withings_data_manger(
```
Please rename to `create_withings_data_manager`.

vangorra committed:
Addressing PR feedback.

vangorra committed:
Formatting source.

MartinHjelmare commented on homeassistant/components/withings/common.py:
```diff
@@ -0,0 +1,308 @@
+)
+
+
+class NotAuthenticatedError(Exception):
```
We should inherit the exceptions we create from `HomeAssistantError`.

MartinHjelmare commented on homeassistant/components/withings/common.py:
```diff
@@ -0,0 +1,308 @@
+
+    async def update_sleep_summary(self):
+        """Update the sleep summary data."""
+        now = datetime.datetime.utcnow()
```
Please use our datetime utils as much as possible. There's `utcnow` in `util/dt.py`.

MartinHjelmare reviewed (commented):
I think we can merge after these comments are addressed.

vangorra committed:
Addressing PR feedback and rebasing.

MartinHjelmare reviewed (approved):
Great!
